### PR TITLE
D3D9: Initial data pitch

### PR DESCRIPTION
**Pull request by K0bin**

### CODE REVIEW

File: src/dxvk/dxvk_util.cpp [hunk 1]
```diff
@@ -71,6 +71,41 @@ namespace dxvk::util {
       }
     }
   }
+
+
+  void packImageData(
```
Comment (Joshua-Ashton) [inline src/dxvk/dxvk_util.cpp:76]: You should make the above packImageData just call this with the info it gets.

**Review by Joshua-Ashton — CHANGES_REQUESTED** [review]

File: src/dxvk/dxvk_util.cpp (hunk 1 above)
Comment (doitsujin) [reply src/dxvk/dxvk_util.cpp:76]: If anything, I'd want everything to use the path below (the one with 8 parameters), if that works. Was too lazy to deal with it yet though.

Comment (Joshua-Ashton) [reply src/dxvk/dxvk_util.cpp:76]: I think having some helper that just calculates and passes it in with less args is fine. Calculating these values everywhere would kinda suck.

Comment (K0bin) [reply src/dxvk/dxvk_util.cpp:63]: I agree with Josh here.

Comment (doitsujin) [reply src/dxvk/dxvk_util.cpp:63]: I'm not really sure what's going on with the d3d9 side but originally these functions were meant to just tightly pack image data so we could use it with `vkCmdCopyBufferToImage`. Looking at what this does in more detail, it's a completely different use case.

Comment (K0bin) [reply src/dxvk/dxvk_util.cpp:63]: We're kinda doing the opposite here. "unpacking" tightly packed data. Would you prefer a separate function for that?

Comment (doitsujin) [reply src/dxvk/dxvk_util.cpp:63]: My problem with this whole thing is that we still have *two* functions doing more or less the thing with different capabilities. I'd rather just have one.

Comment (K0bin) [reply src/dxvk/dxvk_util.cpp:63]: So basically extend the multi-plane variant and use that instead?

File: src/dxvk/dxvk_util.cpp [hunk 2]
```diff
@@ -142,6 +142,90 @@ namespace dxvk::util {
...
+          extent.height /= plane->blockSize.height;
+          elementSize = plane->elementSize;
+        }
+
+        auto blockCount = computeBlockCount(extent, formatInfo->blockSize);
+
+        VkDeviceSize bytesTotal = blockCount.depth * std::min(dstRowPitch, srcRowPitch);
+
+        const bool directCopy = ((dstRowPitch   == srcRowPitch  ) || (blockCount.height == 1))
+                             && ((dstSlicePitch == srcSlicePitch) || (blockCount.depth  == 1));
```
Comment (doitsujin) [inline src/dxvk/dxvk_util.cpp:178]: this is wrong, need to check if both src and dst pitches are equal to the byte count.

File: src/d3d9/d3d9_initializer.cpp [hunk 3]
```diff
@@ -131,18 +132,53 @@ namespace dxvk {
...
+            pitch,
+            pitch * blockCount.height,
+            alignedPitch,
+            alignedPitch * blockCount.height,
+            type,
+            mipExtent,
+            resourceType == D3DRTYPE_CUBETEXTURE ? 6 : 1,
+            formatInfo,
+            VK_IMAGE_ASPECT_COLOR_BIT
+          );
```
Comment (doitsujin) [inline src/d3d9/d3d9_initializer.cpp:176]: what's so important about `);` that it needs its own line?

Comment (K0bin) [reply src/d3d9/d3d9_initializer.cpp:176]: Nothing, I just prefer it like that because it's horizontally aligned with the function call. I'll change that.

Comment (artemyto) [reply src/d3d9/d3d9_initializer.cpp:176]: Kotlin style

Comment (K0bin) [reply src/d3d9/d3d9_initializer.cpp:176]: I like Kotlin.
¯\\\_(ツ)\_/¯

File: src/dxvk/dxvk_util.cpp [hunk 4]
```diff
@@ -142,6 +142,99 @@ namespace dxvk::util {
...
+                dstData + j * dstRowPitch,
+                srcData + j * srcRowPitch,
+                bytesPerRow);
+
+              if (dstRowPitch > bytesPerRow) {
+                std::memset(dstData + j * dstRowPitch + bytesPerRow, 0, dstRowPitch - bytesPerRow);
+              }
+            }
+            if (dstSlicePitch > bytesPerSlice) {
+              std::memset(dstData + bytesPerSlice, 0, dstSlicePitch - bytesPerSlice);
```
Comment (doitsujin) [inline src/dxvk/dxvk_util.cpp:213]: why are we memsetting anything here?

Comment (K0bin) [reply src/dxvk/dxvk_util.cpp:213]: Alright, gonna get rid of that.

File: src/dxvk/dxvk_util.cpp [hunk 5]
```diff
@@ -142,6 +142,99 @@ namespace dxvk::util {
...
+                srcData += srcRowPitch;
+                dstData += dstRowPitch;
+                break;
+              case VK_IMAGE_TYPE_2D:
+                srcData += blockCount.height * srcRowPitch;
+                dstData += blockCount.height * dstRowPitch;
+                break;
+              case VK_IMAGE_TYPE_3D:
+                srcData += srcSlicePitch;
+                dstData += dstRowPitch;
```
Comment (doitsujin) [inline src/dxvk/dxvk_util.cpp:227]: dstSlicePitch

**Comment by Joshua-Ashton** [comment]
Not technically correct given what we know about this host importing the pointer but this is a good stop gap that fixes some games :)

File: src/d3d9/d3d9_initializer.cpp [hunk 6]
```diff
@@ -131,18 +131,41 @@ namespace dxvk {
...
+
+        if (pInitialData != nullptr) {
+          VkExtent3D mipExtent = pTexture->GetExtentMip(m);
+          const DxvkFormatInfo* formatInfo = imageFormatInfo(pTexture->GetFormatMapping().FormatColor);
+          VkExtent3D blockCount = util::computeBlockCount(mipExtent, formatInfo->blockSize);
+          uint32_t pitch = blockCount.width * formatInfo->elementSize;
+          uint32_t alignedPitch = align(pitch, 4);
+
+          D3DRESOURCETYPE resourceType = pTexture->GetType();
+          VkImageType type = resourceType == D3DRTYPE_VOLUME || resourceType == D3DRTYPE_VOLUMETEXTURE
```
Comment (Joshua-Ashton) [inline src/d3d9/d3d9_initializer.cpp:148]: pTexture->GetImage()->info().type

**Comment by K0bin** [comment]
Yup, I'd go with this for 1.9 and do a proper solution in 1.9.1.

Alternatively we can also drop it, given that csgo and Desolation are basically the only games that actually use this as far as I know.

File: src/d3d9/d3d9_initializer.cpp [hunk 7]
```diff
@@ -131,18 +131,41 @@ namespace dxvk {
...
+          util::packImageData(
+            mapSlice.mapPtr,
+            pInitialData,
+            pitch,
+            pitch * blockCount.height,
+            alignedPitch,
+            alignedPitch * blockCount.height,
+            type,
+            mipExtent,
+            resourceType == D3DRTYPE_CUBETEXTURE ? 6 : 1,
```
Comment (Joshua-Ashton) [inline src/d3d9/d3d9_initializer.cpp:161]: `pTexture->Desc()->ArraySize`

**Review by Joshua-Ashton — CHANGES_REQUESTED** [review]

File: src/d3d9/d3d9_initializer.cpp (hunk 6 above)
Comment (K0bin) [reply src/d3d9/d3d9_initializer.cpp:148]: That doesn't work because this is only called for POOL_SYSTEMMEM textures which don't have an image.

**Pull request merged**